libraries/AP_Vehicle/ModeReason.h

```diff
@@ -0,0 +1,54 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <stdint.h>
+
+// interface to set the vehicles mode
+enum class ModeReason : uint8_t {
+  UNKNOWN,
+  RC_COMMAND,
+  GCS_COMMAND,
+  RADIO_FAILSAFE,
+  BATTERY_FAILSAFE,
+  GCS_FAILSAFE,
+  EKF_FAILSAFE,
+  GPS_GLITCH,
+  MISSION_END,
+  THROTTLE_LAND_ESCAPE,
+  FENCE_BREACHED,
+  TERRAIN_FAILSAFE,
+  BRAKE_TIMEOUT,
+  FLIP_COMPLETE,
+  AVOIDANCE,
+  AVOIDANCE_RECOVERY,
+  THROW_COMPLETE,
+  TERMINATE,
+  TOY_MODE,
+  CRASH_FAILSAFE,
+  SOARING_FBW_B_WITH_MOTOR_RUNNING,
+  SOARING_THERMAL_DETECTED,
+  SOARING_THERMAL_ESTIMATE_DETERIORATED,
+  VTOL_FAILED_TRANSITION,
+  VTOL_FAILED_TAKEOFF,
+  FAILSAFE, // general failsafes, prefer specific failsafes over this as much as possible
+  INITIALISED,
+  SURFACE_COMPLETE,
+  BAD_DEPTH,
+  LEAK_FAILSAFE,
+  SERVOTEST,
+  STARTUP,
+  SCRIPTING,
+  UNAVAILABLE,
```

Comment:
what's the difference between UNKNOWN and UNAVAILABLE?

Reply:
Plane uses unavailable for when you tried to change into a quadplane mode, but we don't have quadplane enabled so we bounce you back to a normal/previous plane mode.

Reply:
ok, in copter or rover we would just reject the change and print an error.

Reply:
Yeah, I think we do something like that as well, but the problem with Plane is that we will already have done the exit work on the previous mode and reenter it (and externally/from a switch perspective) mode changes always succeed, even if we then promptly change modes again.